### PR TITLE
Try JIT before falling back to Python.

### PR DESCRIPTION
**Pull request by DrTodd13**
## Changes included in this PR

We do 1) native dataframe library and if that fails now 2) JIT fallback and if that fails 3) Python.

## Testing strategy

run_ci

## User facing changes

None

## Checklist
- [ ] Pipelines passed before requesting review. To run CI you must include `[run CI]` in your commit message.
- [X] I am familiar with the [Contributing Guide](https://github.com/bodo-ai/Bodo/blob/main/CONTRIBUTING.md) 
- [X] I have installed + ran pre-commit hooks.

### CODE REVIEW

File: bodo/pandas/utils.py [hunk 1]
```diff
@@ -1219,3 +1218,129 @@ def silenced_method(*args, **kwargs):
...
+                *[type(x) for x in kwargs.values()],
+            ),
+            None,
+        )
+        # See if we previously tried to compile this function.
+        cache_entry = JITFallback.fallback_cache.get(key, None)
+        if (
+            self.name
+            not in (
+                "items",
```
Comment (scott-routledge2) [inline bodo/pandas/utils.py]: Why does JIT fallback not work for these functions?

File: bodo/pandas/utils.py [hunk 2]
```diff
@@ -1219,3 +1218,129 @@ def silenced_method(*args, **kwargs):
...
+                else:
+                    cm_args = (self.base_obj, *args, *tuple(kwargs.values()))
+                return cache_entry(*cm_args)
+
+        JITFallback.python_fallback += 1
+        msg = (
+            f"{self.name} is not implemented in Bodo Dataframe Library yet. "
+            "Falling back to Pandas (may be slow or run out of memory)."
+        )
+        warnings.warn(BodoLibFallbackWarning(msg))
```
Comment (scott-routledge2) [inline bodo/pandas/utils.py]: This might cause unnecessary warnings like in TPCH `pd.Timestamp("1994-03-01")`. Wondering if there is a way to detect whether a function is performance critical and only warn in that case?

File: bodo/tests/test_df_lib/test_end_to_end.py [hunk 3]
```diff
@@ -325,6 +329,7 @@ def test_filter_pushdown(datapath, file_path, op):
         check_pandas_types=False,
         sort_output=True,
         reset_index=False,
+        check_names=False,
```
Comment (ehsantn) [inline bodo/tests/test_df_lib/test_end_to_end.py]: Why is this changed? How would this PR impact output names?

**Review by ehsantn — APPROVED** [review]
Thanks @DrTodd13! Looks good to me overall. I think some corner cases may pop up from time to time that we just need to handle later.

File: bodo/tests/test_df_lib/test_end_to_end.py (hunk 3 above)
Comment (DrTodd13) [reply bodo/tests/test_df_lib/test_end_to_end.py]: Was necessary with previous version of code but not the re-worked code so I've reverted this.

File: bodo/pandas/utils.py [hunk 4]
```diff
@@ -1232,3 +1239,102 @@ def silenced_method(*args, **kwargs):
...
+                if self.base_obj is None:
+                    cm_args = args + tuple(kwargs.values())
+                else:
+                    cm_args = (self.base_obj, *args, *tuple(kwargs.values()))
+                return cache_entry(*cm_args)
+
+        JITFallback.python_fallback += 1
+        raise JITFallback.JITFallbackFail()
+
+        # msg = (
```
Comment (ehsantn) [inline bodo/pandas/utils.py]: Remove commented out code?

File: bodo/pandas/utils.py [hunk 5]
```diff
@@ -1232,3 +1239,102 @@ def silenced_method(*args, **kwargs):
...
+                        [f"arg{i}" for i in range(len(args))]
+                        + [f"{x}={x}" for x in kwargs.keys()]
+                    )
+                    func_text = f"def {fname}({self_arg}{sig_args}):\n"
+                    if self.base_obj is None:
+                        func_text += f"    return pd.{self.name}({caller_args})\n"
+                    else:
+                        func_text += f"    return self.{self.name}({caller_args})\n"
+
+                    # print("new_func_text:", func_text)
```
Comment (ehsantn) [inline bodo/pandas/utils.py]: Remove

File: bodo/pandas/utils.py [hunk 6]
```diff
@@ -1232,3 +1239,102 @@ def silenced_method(*args, **kwargs):
...
+                self.name,
+                *[type(x) for x in args],
+                *list(kwargs.keys()),
+                *[type(x) for x in kwargs.values()],
+            ),
+            None,
+        )
+        # See if we previously tried to compile this function.
+        cache_entry = JITFallback.fallback_cache.get(key, None)
+        if self.name in ("duplicated", "pivot") and cache_entry != False:
```
Comment (ehsantn) [inline bodo/pandas/utils.py:1280]: Is `self.name in ("duplicated", "pivot")` debug code that needs removed? Or is it related to unit testing?

**Review by ehsantn — APPROVED** [review]
Looks good to me. Thanks @DrTodd13

File: bodo/pandas/utils.py (hunk 4 above)
Comment (DrTodd13) [reply bodo/pandas/utils.py]: done

File: bodo/pandas/utils.py (hunk 5 above)
Comment (DrTodd13) [reply bodo/pandas/utils.py]: done

File: bodo/pandas/utils.py (hunk 6 above)
Comment (DrTodd13) [reply bodo/pandas/utils.py:1280]: This is the whitelist I mentioned.  Scott had a good point about too many warnings and performance critical functions so I turned off the warning for non-callable attributes of the class.  There are some routines for which the JIT fallback function runs but then the output of that is of some form that breaks things downstream.  We don't know what we don't know so even if I excluded things I don't think work then there are surely untested things that also don't work which will break our claims of seamless auto-fallback.  We can revisit later.

Comment (ehsantn) [reply bodo/pandas/utils.py:1280]: Makes sense. Let's just add some comments in the code and make the list explicit.

**Review by scott-routledge2 — APPROVED** [review]
LGTM, thanks @DrTodd13
